capsules/src/lpm013m126.rs

```diff
@@ -0,0 +1,487 @@
+        frame_buffer.data
+    }
+}
+
+/// Modes are 6-bit, network order.
+/// They use a tree-ish encoding, so only the ones in use are listed here.
+#[derive(Clone, Copy)]
+enum Mode {
+    /// Clear memory
+    /// bits: 0 Function, X, 1 Clear, 0 Blink off, X, X
+    AllClear = 0b001000,
+    /// Invert colors periodically
+    /// bits: 0 Function, X, 0 No Clear, 1 Blink, 1 White, 1 Invert
+    BlinkInversion = 0b000111,
+    /// Input 1-bit data
+    /// bits: 1 No function, X, 0 Data Update, 01 1-bit, X
+    Input1Bit = 0b100_01_0,
+}
+
+/// Command header is composed of a 6-bit mode and 10 bits of address,
+/// network bit order.
+struct CommandHeader {
+    mode: Mode,
+    gate_line: u16,
+}
+
+impl CommandHeader {
+    /// Formats header for transfer
+    fn encode(&self) -> [u8; 2] {
+        (self.gate_line | ((self.mode as u16) << 10)).to_be_bytes()
+    }
+}
+
+type WritePosition = WriteFrame;
+
+/// Area of the screen to which data is written
+#[derive(Debug, Copy, Clone)]
+struct WriteFrame {
+    row: u16,
+    column: u16,
+    width: u16,
+    height: u16,
+}
+
+/// Position within the write frame, rows first.
+type Position = u16;
+
+/// Internal state of the driver.
+/// Each state can lead to the next one in order of appearance.
+#[derive(Debug, Copy, Clone)]
+enum State {
+    Uninitialized,
+    InitializingPixelMemory,
+    /// COM polarity and internal latch circuits
+    InitializingRest,
+
+    // Normal operation
+    Idle(WriteFrame),
+    Writing(WriteFrame),
+
+    /// This driver is buggy. Resetting will try to recover it.
+    Bug,
+}
+
+pub struct Lpm013m126<'a, A: Alarm<'a>, P: Pin, S: SpiMasterDevice> {
+    spi: &'a S,
+    extcomin: &'a P,
+    disp: &'a P,
+
+    state: Cell<State>,
+
+    /// The HIL requires updates to arbitrary rectangles.
+    /// The display supports only updating entire rows,
+    /// so edges need to be cached.
+    frame_buffer: OptionalCell<FrameBuffer<'static>>,
+
+    client: OptionalCell<&'static dyn ScreenClient>,
+    /// Buffer for incoming pixel data.
+    /// It's not submitted directly anywhere.
+    buffer: TakeCell<'static, [u8]>,
+
+    /// Needed forinit and to flip the EXTCOMIN pin at regular intervals
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>, P: Pin, S: SpiMasterDevice> Lpm013m126<'a, A, P, S> {
+    /// Caution: passing `frame_buffer` that's too small will panic.
+    pub fn new(
+        spi: &'a S,
+        extcomin: &'a P,
+        disp: &'a P,
+        alarm: &'a A,
+        frame_buffer: &'static mut [u8],
+    ) -> Self {
+        Self {
+            spi,
+            alarm: alarm,
+            disp,
+            extcomin,
+            frame_buffer: OptionalCell::new(FrameBuffer::new(frame_buffer)),
+            buffer: TakeCell::empty(),
+            client: OptionalCell::empty(),
+            state: Cell::new(State::Uninitialized),
+        }
+    }
+
+    pub fn initialize(&self) -> Result<(), ErrorCode> {
```

Comment:
Can you add a comment explaining when initialize should be called? It appears that initialize can be safely called during the board's init process, even if the screen is never used, but it would be good to clarify.

Reply:
Instead of adding a comment here, I added it to setup, and un-pub'ed initialize.